### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by seletskiy**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,42 +3,525 @@ tests.sh
...
+    tests.sh [-v] [-d <dir>] -O [<name>]
+    tests.sh -i
 
 Options:
     -h | --help  Show this help.
     -A           Run all testcases in current directory.
-    -O <name>    Run specified testcase only.
+    -O <name>    Run specified testcase only. If no testcase specified, last failed
+                 testcase will be ran.
+    -d <dir>     Change directory to specified before running testcases.
```
Comment (kovetskiy) [inline README.md:30]: without default value?

File: README.md [hunk 2]
```diff
@@ -3,42 +3,525 @@ tests.sh
...
+
+Local set up script should be named `local.setup.sh` and be located in the
+same directory as testcases. It will be sources every time before each
+testcase.
+
+# Reference
+
+## tests:import-namespace()
+
+Make all functions from tests.sh available without 'tests:'
```
Comment (kovetskiy) [inline README.md:63]: extra line break

File: tests.sh [hunk 3]
```diff
@@ -1,78 +1,105 @@
...
-#   Path to temp dir, e.g, /tmp/tests.XXXX
-tests_tmpdir() {
-    if [[ "$TEST_DIR" == "" ]]; then
-        tests_debug "test session not initialized"
+# @description Make all functions from tests.sh available without 'tests:'
+# prefix.
+#
+# @noargs
+tests:import-namespace() {
+    builtin eval $(declare -F | grep -F -- '-f tests:' | cut -d: -f2 |
```
Comment (kovetskiy) [inline tests.sh]: I think we should not use admin-like bash style and write code with pipes using
common programming languages important thing (omitting corner-cases):

##### one line - one operation

so, I want to say we should write bash code with pipes using
extra line breaks like as following:

``` bash
tests:import-namespace() {
    builtin eval $(
        declare -F \
            | grep -F -- '-f tests:' \
            | cut -d: -f2 \
            | sed -re's/.*/&() { tests:& "${@}"; };/'
    )
}
```

File: tests.sh [hunk 4]
```diff
@@ -1,78 +1,105 @@
...
+        sed -re's/.*/&() { tests:& "${@}"; };/'
+    )
+}
+
+# @description Returns temporary directory for current test session.
```
Comment (kovetskiy) [inline tests.sh:25]: `@description` ? why not like in Go? take a doc-block before a function, remove special blocks (`@example`, `@echo`, `@args`), trim and use leftover as description.

File: tests.sh [hunk 5]
```diff
@@ -1,78 +1,105 @@
...
+#
+# It can be used as a workspace for the testcase.
+#
+# @example
+#   ls $(tests:get-tmp-dir)
+#
+# @echo Path to temp dir, e.g, /tmp/tests.XXXX
```
Comment (kovetskiy) [inline tests.sh]: hmm, may be `@output` instead of `@echo`? I think doc tags should be nouns, not verbs.

File: tests.sh [hunk 6]
```diff
@@ -1,78 +1,105 @@
...
+tests:get-tmp-dir() {
+    if [[ "$tests_dir" == "" ]]; then
```
Comment (kovetskiy) [inline tests.sh]: [ -n "$tests_dir" ]

File: tests.sh [hunk 7]
```diff
@@ -1,78 +1,105 @@
...
-# Return:
-#   1: if strings is not equals
-#   0: otherwise
-tests_assert_equals() {
+# @description Asserts, that first string arg is equals to second.
+#
+# @example
+#   tests:assert-equals 1 2 # fails
+#
+# @arg $1 string Expected string.
```
Comment (kovetskiy) [inline tests.sh:47]: Can we beat this `$1` `$2` options and remove they from doc?

File: tests.sh [hunk 8]
```diff
@@ -80,58 +107,60 @@ tests_re() {
...
     return $?
 }
 
-# Function tests_assert_re checks that last evaluated command output
-# (stdout or stderr) matches regexp.
-# Args:
-#   $1: stdout|stderr|<filename>
-#   $2: regexp
-tests_assert_re() {
+# @description Same, as 'tests:match-re', but abort testing if comparison failed.
```
Comment (kovetskiy) [inline tests.sh]: s/Same,/Same

File: tests.sh [hunk 9]
```diff
@@ -80,58 +107,60 @@ tests_re() {
...
-        tests_debug ">>> ${regexp:-<empty regexp>}"
-        tests_debug "<<< ${target}"
-        tests_debug "<<< $(cat $file)"
+    if ! tests_make_assertion $result 0; then
+        touch "$tests_dir/.failed"
+        tests:debug "expectation failed: regexp does not match"
+        tests:debug "*** match on: ${target}"
+        tests:debug ">>> ${regexp:-<empty regexp>}"
+        tests:debug "<<<"
+        cat $file | tests_indent
```
Comment (kovetskiy) [inline tests.sh]:
```
 λ  ~ echo xxx > /tmp/file_xxx
 λ  ~ (echo 1; cat; echo 2) < /tmp/file_xxx
1
xxx
2
```

File: tests.sh [hunk 10]
```diff
@@ -80,58 +107,60 @@ tests_re() {
...
     local result=$?
 
-    if [ $result -gt 0 ]; then
-        touch "$TEST_DIR/_failed"
-        tests_debug "expectation failed: regexp does not match"
-        tests_debug ">>> ${regexp:-<empty regexp>}"
-        tests_debug "<<< ${target}"
-        tests_debug "<<< $(cat $file)"
+    if ! tests_make_assertion $result 0; then
+        touch "$tests_dir/.failed"
```
Comment (kovetskiy) [inline tests.sh]: please create additional function for this, like tests:fail

File: tests.sh [hunk 11]
```diff
@@ -169,187 +198,329 @@ tests_diff() {
...
-tests_diff_blank() {
-    tests_diff "$1" "$2" "-B"
+# @description Returns file containing stdout of last command.
+#
+# @example
+#   tests:eval echo 123
+#   cat $(tests:get-stdout) # will echo 123
+#
+# @echo Filename containing stdout.
+tests:get-stdout() {
```
Comment (kovetskiy) [inline tests.sh]: weird name, returns file with stdout instead of output

guess should create `tests:stdout()` for contents of stdout

File: tests.sh [hunk 12]
```diff
@@ -169,187 +198,329 @@ tests_diff() {
...
-# Args:
-#   $@: arguments for 'test' utility
-tests_test() {
+# @description Same as 'tests:assert-diff', but ignore changes whose lines are
+# all blank.
+#
+# @example
+#   tests:eval echo -e '1\n2'
+#   tests:assert-no-diff stdout "$(echo -e '1\n2')" # note quotes
+#   tests:assert-no-diff stdout "$(echo -e '1\n\n2')" # test will pass
```
Comment (kovetskiy) [inline tests.sh]: s/diff/diff-blank/

File: tests.sh [hunk 13]
```diff
@@ -169,187 +198,329 @@ tests_diff() {
...
+#
+# @see tests:diff
+tests:assert-no-diff-blank() {
+    tests:assert-no-diff "$1" "$2" "-B"
+}
+
+# @description Same, as shell 'test' function, but asserts, that exit code is
```
Comment (kovetskiy) [inline tests.sh]: s/Same,/Same/g

File: tests.sh [hunk 14]
```diff
@@ -169,187 +198,329 @@ tests_diff() {
...
+# @arg $1 filename Temporary file name.
+# @arg $2 string Contents to put.
+tests:put-string() {
     local file="$1"
     local content="$2"
-    tests_put "$TEST_DIR/$file" "$content"
+
+    tests:debug "writing a file $file with content:"
+    echo "$content" | tests_indent
+
```
Comment (kovetskiy) [inline tests.sh]: `echo -n` may be?

File: tests.sh [hunk 15]
```diff
@@ -169,187 +198,329 @@ tests_diff() {
...
+#
+# @example
+#   tests:eval false
+#   tests:assert-exitcode 1
+#
+# @arg $1 int Expected exit code.
+tests:assert-exitcode() {
+    local actual=$?
+    local expected=$1
     shift
```
Comment (kovetskiy) [inline tests.sh:382]: looks strange, please move `actual=$?` below than `shift`

File: tests.sh [hunk 16]
```diff
@@ -169,187 +198,329 @@ tests_diff() {
...
+    builtin eval "${@}"
+    tests_assert_operation="="
+}
+
+# @description Evaluates given command and show it's output in the debug, used
+# for debug purposes.
+#
+# @arg $@ any Command to evaluate.
+tests:describe() {
+    tests:debug "this test decription:"
```
Comment (kovetskiy) [inline tests.sh]: better to use colors or header insead

File: tests.sh [hunk 17]
```diff
@@ -169,187 +198,329 @@ tests_diff() {
...
-tests_cd() {
+# @description Changes working directory to specified directory.
+#
+# @arg $1 directory Directory to change to.
+tests:cd() {
     local dir=$1
-    tests_debug "\$ cd $1"
-    cd $1
+    tests:debug "\$ cd $1"
+    builtin cd $1
```
Comment (kovetskiy) [inline tests.sh:446]: $dir

File: tests.sh [hunk 18]
```diff
@@ -169,187 +198,329 @@ tests_diff() {
...
+tests:ensure() {
+    tests:eval "$@"
+    tests:assert-success
 }
 
-tests_mkdir() {
-    tests_do mkdir -p $TEST_DIR/$1
+# @description Creates temporary directory.
+#
+# @arg $1 string Directory name.
```
Comment (kovetskiy) [inline tests.sh]: $@

Comment (kovetskiy) [inline tests.sh]: should use mkdir -p

File: tests.sh [hunk 19]
```diff
@@ -169,187 +198,329 @@ tests_diff() {
...
+tests:mkdir() {
+    # prepend to any non-flag argument $tests_dir prefix
+    tests:eval /bin/mkdir \
+        $(sed -re "s#(^|\\s)([^-])#\\1$tests_dir/\\2#g" <<< "${@}")
```
Comment (kovetskiy) [inline tests.sh]: add quotes

File: tests.sh [hunk 20]
```diff
@@ -435,9 +621,87 @@ tests_wait_file_changes() {
...
+
+# @description Recursively copy specified file or directory from the testcases
+# dir to the temporary test directory.
+#
+# @arg $1 filename Filename to copy.
+tests:cp() {
+    local files=($tests_case_dir/$1)
+    shift
+
+    while [ $# -gt 1 ]; do
```
Comment (kovetskiy) [inline tests.sh]: strange condition `argv > 1`, can you explain?

File: tests.sh [hunk 21]
```diff
@@ -435,9 +621,87 @@ tests_wait_file_changes() {
...
+}
+
+# @description Copy specified file from testcases to the temporary test
+# directory and then source it.
+#
+# @arg $1 filename Filename to copy and source.
+tests:source() {
+    tests:cp "$1"
+
+    local source_name=$(basename $1)
```
Comment (kovetskiy) [inline tests.sh]: add local file="$1"

File: tests.sh [hunk 22]
```diff
@@ -435,9 +621,87 @@ tests_wait_file_changes() {
...
+    tests:debug "{BEGIN} source $source_name"
+    builtin source "$source_name"
+    tests:debug "{END} source $source_name"
```
Comment (kovetskiy) [inline tests.sh]: strance construction "{BEGIN}" "{END}", better to write more obvious debug messages like just `source $source_name` and `$source_name sourced`

File: tests.sh [hunk 23]
```diff
@@ -652,36 +948,184 @@ tests_interrupt() {
...
-    local count=$(cat $TEST_DIR/_asserts)
-    echo $(($count+1)) > $TEST_DIR/_asserts
+    local count=$(cat $tests_dir/.asserts)
+    echo $(($count+1)) > $tests_dir/.asserts
 }
 
-tests_source() {
-    tests_copy "$1"
+tests_print_docs() {
+    parser='# start of awk code {{{
```
Comment (kovetskiy) [inline tests.sh:1191]: move it to docs.awk

**Comment by kovetskiy** [comment]
Too much breaking changes, can you create a migrating tool?

File: README.md (hunk 2 above)
Comment (seletskiy) [inline README.md:63]: Why it's extra? README.md seems to render correctly.

Comment (kovetskiy) [inline README.md:63]: because this line length lower than 80

File: tests.sh (hunk 4 above)
Comment (seletskiy) [inline tests.sh:25]: Unfortunately, bash does not have named/typed function arguments, so golang-style documentation will lead to sentences like "$1 is blah blah. $2 is blah blah", which is somewhat difficult for reading.

File: tests.sh (hunk 5 above)
Comment (seletskiy) [inline tests.sh]: Agree, but `@stdout` is preferrable instead of `@output`.

File: tests.sh (hunk 9 above)
Comment (seletskiy) [inline tests.sh]: What do you mean?

Comment (kovetskiy) [inline tests.sh]: you can use just `tests_indent < $file` without using cat and pipe

File: README.md (hunk 2 above)
Comment (seletskiy) [inline README.md:63]: It's autogenerated from shdoc, so in the file that line fits 80 nicely.
